text/0004-import-maps.md

```diff
@@ -0,0 +1,82 @@
+# Import maps
+- Start Date: 2019/05/01
+- RFC PR: https://github.com/openmrs/openmrs-rfc-frontend/pull/4
+
+## Decision, including impact on distributions
+Within the [OpenMRS master single page application](/text/0001-single-page-applications.md), we use import maps to control which url each
+of the [in-browser javascript modules](/text/0002-modules.md) should be downloaded from. Since import maps are not supported by all browsers yet,
+we use SystemJS to polyfill import maps.
+
+All in-browser OpenMRS core modules and their URLs are defined in an import map. Distributions are encouraged to add their in-browser modules to an import map, too.
+
+## Definition
+An [import map](https://github.com/WICG/import-maps) is a browser specification for controlling the url to download in-browser javascript modules from. At the
+time of this writing, import maps are currently a proposal that has not yet been voted on by major browsers and standards bodies. However, Chrome has already
+implemented import maps behind a developer feature flag. The [WICG](https://wicg.github.io/admin/charter.html) is the organization backing and proposing import
+maps, with Domenic Denicola being the chief author of the proposal.
+
+To use browser import maps, you add a `<script>` element to your html page:
+```html
+<html>
+  <head>
+    <script type="importmap">
+      {
+        "imports": {
+          "@openmrs/styleguide": "http://localhost:8080/styleguide.js"
+        }
+      }
+    </script>
+
+    <!-- The browser understands the unusual src attribute because of the import map we have defined above -->
+    <script type="module" src="import:@openmrs/styleguide"></script>
+  </head>
+  <body>
+  </body>
+</html>
+```
+
+Because import maps are not implemented in browsers yet, we use [SystemJS](https://github.com/systemjs/systemjs) as a polyfill for import maps.
+The syntax for SystemJS import maps is slightly different than the specification:
```

Comment:
Any reason for making it different from the specification?

Reply:
The only difference is changing the `<script type="importmap">` to `<script type="systemjs-importmap">`. The inner json is exactly the same.

I believe the reason that it is different is so that the developer has full control over whether they are using native browser import maps vs systemjs import maps. Using the exact same script type could result in users accidentally using native browser import maps instead of SystemJS if they update their browser to a version that uses native import maps. Since the specification for import maps is still slightly evolving, it is safer to stick to SystemJS until we're confident that our code will work in all browsers' implementation of import maps.

^^ This is my best guess as to the explanation, but if you want to ask the author of SystemJS, you can open an issue in https://github.com/systemjs/systemjs asking about it. I am a contributor to the project, but not the primary author and maintainer.

Reply:
Oh i see! Got you. 😊